File: frontend/src/api/shop.ts
import { apiFetch } from './client';
import type { Diet, MenuItem, Order, PrepItem, ShopState, ShopStatus } from './types';

export interface MenuItemInput {
  name: string;
  price: number; // paise
  photo_url: string | null;
  avail_from: string | null;
  avail_to: string | null;
  is_available: boolean;
  // The backend REQUIRES diet on create/update (400 otherwise); these are
  // optional in TS only so the not-yet-updated menu form keeps compiling.
  // WP2's menu form must always send diet, and tags as a (possibly empty) array.
  diet?: Diet;
  tags?: string[];
}

export async function getShopMenu(): Promise<MenuItem[]> {
  const res = await apiFetch<{ items: MenuItem[] }>('/shop/menu');
  return res.items;
}

export async function createMenuItem(input: MenuItemInput): Promise<MenuItem> {
  const res = await apiFetch<{ item: MenuItem }>('/shop/menu', {
    method: 'POST',
    body: input,
  });
  return res.item;
}

export async function updateMenuItem(id: number, input: MenuItemInput): Promise<MenuItem> {
  const res = await apiFetch<{ item: MenuItem }>(`/shop/menu/${id}`, {
    method: 'PUT',
    body: input,
  });
  return res.item;
}

export async function deleteMenuItem(id: number): Promise<void> {
  await apiFetch<void>(`/shop/menu/${id}`, { method: 'DELETE' });
}

export async function setMenuItemStock(id: number, outOfStock: boolean): Promise<MenuItem> {
  const res = await apiFetch<{ item: MenuItem }>(`/shop/menu/${id}/stock`, {
    method: 'POST',
    body: { out_of_stock: outOfStock },
  });
  return res.item;
}

export interface ShopOrders {
  incoming: Order[];
  in_progress: Order[];
  awaiting_payment: Order[];
}

export async function getShopOrders(): Promise<ShopOrders> {
  return apiFetch<ShopOrders>('/shop/orders');
}

export interface HistoryItemCount {
  name: string;
  qty: number;
}

export interface HistoryCustomer {
  name: string;
  order_count: number;
}

/** Day summary computed over completed (paid) orders. */
export interface HistoryInsights {
  order_count: number;
  item_counts: HistoryItemCount[]; // sorted qty desc
  customers: HistoryCustomer[]; // sorted order_count desc
}

export interface ShopHistory {
  orders: Order[];
  total_paid: number; // paise collected today
  insights: HistoryInsights;
}

/** Today's finished orders + paid total + insights, for counter reconciliation. */
export async function getShopHistory(date?: string): Promise<ShopHistory> {
  const query = date ? `?date=${date}` : '';
  return apiFetch<ShopHistory>(`/shop/history${query}`);
}

/** Public: the canteen's current open/paused/closed status. */
export async function getShopStatus(): Promise<ShopStatus> {
  return apiFetch<ShopStatus>('/shop-status');
}

/**
 * Shopkeeper: set the canteen status. Pass reopenAt (RFC3339) only when pausing.
 * The backend returns 409 ("Finish or cancel the N active order(s) first.") if
 * pausing/closing while any order is still active.
 */
export async function setShopStatus(state: ShopState, reopenAt?: string | null): Promise<ShopStatus> {
  return apiFetch<ShopStatus>('/shop/status', {
    method: 'POST',
    body: { state, reopen_at: reopenAt ?? null },
  });
}

export async function acceptOrder(id: number, rejectedItemIds: number[]): Promise<Order> {
  const res = await apiFetch<{ order: Order }>(`/shop/orders/${id}/accept`, {
    method: 'POST',
    body: { rejected_item_ids: rejectedItemIds },
  });
  return res.order;
}

export async function rejectOrder(id: number): Promise<Order> {
  const res = await apiFetch<{ order: Order }>(`/shop/orders/${id}/reject`, {
    method: 'POST',
  });
  return res.order;
}

export async function handoverItem(orderId: number, itemId: number, qty = 1): Promise<Order> {
  const res = await apiFetch<{ order: Order }>(`/shop/orders/${orderId}/items/${itemId}/handover`, {
    method: 'POST',
    body: { qty },
  });
  return res.order;
}

/** Remove a line from an accepted order; prepared units return to the pool. */
export async function removeOrderItem(orderId: number, itemId: number): Promise<Order> {
  const res = await apiFetch<{ order: Order }>(`/shop/orders/${orderId}/items/${itemId}`, {
    method: 'DELETE',
  });
  return res.order;
}

export async function markPaid(orderId: number): Promise<Order> {
  const res = await apiFetch<{ order: Order }>(`/shop/orders/${orderId}/paid`, {
    method: 'POST',
  });
  return res.order;
}

export async function getPrep(): Promise<PrepItem[]> {
  const res = await apiFetch<{ items: PrepItem[] }>('/shop/prep');
  return res.items;
}

export async function markPrepDone(menuItemId: number, qty = 1): Promise<void> {
  await apiFetch<{ ok: boolean }>(`/shop/prep/${menuItemId}/done`, {
    method: 'POST',
    body: { qty },
  });
}

export interface PhotoSignature {
  signature: string;
  timestamp: number;
  api_key: string;
  cloud_name: string;
  folder: string;
}

export async function getPhotoUploadSignature(): Promise<PhotoSignature> {
  return apiFetch<PhotoSignature>('/shop/menu/photo-signature', { method: 'POST' });
}

/** Uploads a menu item photo directly to Cloudinary (signed upload — the API
 * secret never reaches the browser) and returns the resulting secure_url. */
export async function uploadMenuItemPhoto(file: File): Promise<string> {
  const sig = await getPhotoUploadSignature();

  const formData = new FormData();
  formData.append('file', file);
  formData.append('api_key', sig.api_key);
  formData.append('timestamp', sig.timestamp.toString());
  formData.append('signature', sig.signature);
  formData.append('folder', sig.folder);

  const res = await fetch(`https://api.cloudinary.com/v1_1/${sig.cloud_name}/image/upload`, {
    method: 'POST',
    body: formData,
  });

  if (!res.ok) {
    throw new Error('Failed to upload photo');
  }

  const data = await res.json();
  return data.secure_url;
}

export async function getVapidPublicKey(): Promise<{ public_key: string }> {
  return apiFetch<{ public_key: string }>('/push/vapid-public-key');
}

export async function subscribeToPush(endpoint: string, p256dh: string, auth: string): Promise<void> {
  await apiFetch('/push/subscribe', {
    method: 'POST',
    body: { endpoint, keys: { p256dh, auth } },
  });
}

File: backend/internal/controllers/shop.go
package controllers

import (
	"net/http"
	"time"

	"github.com/gin-gonic/gin"

	"khaao/internal/config"
	"khaao/internal/middleware"
	"khaao/internal/models"
	"khaao/internal/realtime"
	"khaao/internal/services"
)

type ShopController struct {
	orderService *services.OrderService
	poolEngine   *services.PoolEngine
	hub          *realtime.Hub
	cfg          *config.Config
}

func NewShopController(os *services.OrderService, pe *services.PoolEngine, hub *realtime.Hub, cfg *config.Config) *ShopController {
	return &ShopController{orderService: os, poolEngine: pe, hub: hub, cfg: cfg}
}

func (sc *ShopController) Orders(c *gin.Context) {
	incoming, active, awaiting, err := sc.orderService.ShopOrders(c.Request.Context())
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, gin.H{"incoming": incoming, "in_progress": active, "awaiting_payment": awaiting})
}

type acceptRequest struct {
	RejectedItemIDs []uint `json:"rejected_item_ids"`
}

func (sc *ShopController) Accept(c *gin.Context) {
	orderID, err := parseUintParam(c, "id")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid id"})
		return
	}
	var req acceptRequest
	if c.Request.ContentLength != 0 {
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "invalid request body"})
			return
		}
	}
	order, err := sc.poolEngine.Accept(c.Request.Context(), orderID, req.RejectedItemIDs)
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, gin.H{"order": order})
}

func (sc *ShopController) Reject(c *gin.Context) {
	orderID, err := parseUintParam(c, "id")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid id"})
		return
	}
	order, err := sc.poolEngine.Reject(c.Request.Context(), orderID)
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, gin.H{"order": order})
}

func (sc *ShopController) Prep(c *gin.Context) {
	items, err := sc.poolEngine.PrepList(c.Request.Context())
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, gin.H{"items": items})
}

type doneRequest struct {
	Qty int `json:"qty"`
}

func (sc *ShopController) Done(c *gin.Context) {
	menuItemID, err := parseUintParam(c, "menu_item_id")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid menu item id"})
		return
	}
	req := doneRequest{Qty: 1}
	if c.Request.ContentLength != 0 {
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "invalid request body"})
			return
		}
	}
	if err := sc.poolEngine.MarkDone(c.Request.Context(), menuItemID, req.Qty); err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, gin.H{"ok": true})
}

type handoverRequest struct {
	Qty int `json:"qty"`
}

func (sc *ShopController) Handover(c *gin.Context) {
	orderID, err := parseUintParam(c, "id")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid id"})
		return
	}
	itemID, err := parseUintParam(c, "itemID")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid item id"})
		return
	}
	req := handoverRequest{Qty: 1}
	if c.Request.ContentLength != 0 {
		if err := c.ShouldBindJSON(&req); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "invalid request body"})
			return
		}
	}
	order, err := sc.poolEngine.Handover(c.Request.Context(), orderID, itemID, req.Qty)
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, gin.H{"order": order})
}

func (sc *ShopController) RemoveItem(c *gin.Context) {
	orderID, err := parseUintParam(c, "id")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid id"})
		return
	}
	itemID, err := parseUintParam(c, "itemID")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid item id"})
		return
	}
	order, err := sc.poolEngine.RemoveItem(c.Request.Context(), orderID, itemID)
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, gin.H{"order": order})
}

func (sc *ShopController) Paid(c *gin.Context) {
	orderID, err := parseUintParam(c, "id")
	if err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "invalid id"})
		return
	}
	order, err := sc.poolEngine.Paid(c.Request.Context(), orderID)
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, gin.H{"order": order})
}

func (sc *ShopController) History(c *gin.Context) {
	date := c.Query("date")
	if date == "" {
		date = models.DayOf(time.Now().In(sc.cfg.Location()))
	} else if _, err := time.Parse("2006-01-02", date); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "date must be YYYY-MM-DD"})
		return
	}
	orders, totalPaid, insights, err := sc.orderService.ShopHistory(c.Request.Context(), date)
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, gin.H{"orders": orders, "total_paid": totalPaid, "insights": insights})
}

func (sc *ShopController) Stream(c *gin.Context) {
	streamSSE(c, sc.hub, middleware.UserID(c), "shopkeeper")
}
